src/app/loadout-builder/filter/LoadoutOptimizerMenuItems.tsx

```diff
@@ -230,61 +95,221 @@ export default memo(function LockArmorAndPerks({
+      </div>
       {showExoticPicker && (
         <Portal>
           <ExoticPicker
             lockedExoticHash={lockedExoticHash}
-            classType={selectedStore.classType}
+            classType={classType}
             onSelected={(exotic) => lbDispatch({ type: 'lockExotic', lockedExoticHash: exotic })}
             onClose={() => setShowExoticPicker(false)}
           />
         </Portal>
       )}
-      {showSubclassOptionsPicker && subclass && (
-        <Portal>
-          <SubclassPlugDrawer
-            subclass={subclass.item}
-            socketOverrides={subclass.loadoutItem.socketOverrides ?? emptyObject()}
-            onAccept={(socketOverrides) =>
-              lbDispatch({ type: 'updateSubclassSocketOverrides', socketOverrides, subclass })
-            }
-            onClose={() => setShowSubclassOptionsPicker(false)}
-          />
-        </Portal>
-      )}
     </>
   );
 });
+
+export const LoadoutOptimizerPinnedItems = memo(function LoadoutOptimizerPinnedItems({
+  chooseItem,
+  selectedStore,
+  pinnedItems,
+  searchFilter,
+  lbDispatch,
+}: {
+  chooseItem: ChooseItemFunction;
+  selectedStore: DimStore;
+  pinnedItems: PinnedItems;
+  searchFilter: ItemFilter;
+  lbDispatch: Dispatch<LoadoutBuilderAction>;
+}) {
+  /**
+   * Lock currently equipped items on a character
+   * Recomputes matched sets
+   */
+  const lockEquipped = () =>
+    lbDispatch({
+      type: 'setPinnedItems',
+      items: selectedStore.items.filter((item) => item.equipped && isLoadoutBuilderItem(item)),
+    });
+
+  const pinItem = useCallback(
+    (item: DimItem) => lbDispatch({ type: 'pinItem', item }),
+    [lbDispatch]
+  );
+  const unpinItem = (item: DimItem) => lbDispatch({ type: 'unpinItem', item });
+
+  const chooseLockItem = chooseItem(
+    pinItem,
+    // Exclude types that already have a locked item represented
+    (item) => Boolean(!pinnedItems[item.bucket.hash] && searchFilter(item))
+  );
+
+  const allPinnedItems = _.sortBy(_.compact(objectValues(pinnedItems)), (i) =>
+    LockableBucketHashes.indexOf(i.bucket.hash)
+  );
+
+  return (
+    <LoadoutBucketDropTarget className={styles.area} onItemLocked={pinItem}>
+      {Boolean(allPinnedItems.length) && (
+        <div className={styles.itemGrid}>
+          {allPinnedItems.map((lockedItem) => (
+            <LockedItem key={lockedItem.id} lockedItem={lockedItem} onRemove={unpinItem} />
+          ))}
+        </div>
+      )}
+      <div className={styles.buttons}>
+        <button type="button" className="dim-button" onClick={chooseLockItem}>
+          <AppIcon icon={pinIcon} /> {t('LoadoutBuilder.LockItem')}
+        </button>
+        <button type="button" className="dim-button" onClick={lockEquipped}>
+          <AppIcon icon={pinIcon} /> {t('LoadoutBuilder.LockEquipped')}
+        </button>
+      </div>
+    </LoadoutBucketDropTarget>
+  );
+});
+
+export const LoadoutOptimizerExcludedItems = memo(function LoadoutOptimizerExcludedItems({
```

Comment:
I didn't look into all the components split out but I am not sure this `memo` adds any value. If the parent rerenders, the `chooseItem` function is unstable so will trigger a rerender anyways.

Reply:
Got it - I'll add useCallback to that. I kept them as memo just because the original huge component had it.